proposals/csharp-10.0/parameterless-struct-constructors.md

```diff
@@ -0,0 +1,106 @@
+# Parameterless struct constructors
+
+## Summary
+
+Support explicit parameterless constructors for value types.
+
+## Proposal
+
+If the parameterless constructor is less accessible than the value type, a warning is reported and the constructor is ignored when inaccessible.
+```csharp
+struct NoConstructor
+{
+    // ok
+}
+struct PublicConstructor
+{
+    internal PublicConstructor() { } // ok: always accessible
+}
+struct PrivateConstructor
+{
+    private PrivateConstructor() { } // warning: parameterless .ctor will be ignored if inaccessible
+}
+```
+
+`default(S)` ignores any parameterless constructor and generates a zeroed `S`.
+```csharp
+NoConstructor x = default;      // ok
+PublicConstructor y = default;  // ok
+PrivateConstructor z = default; // ok
+```
+
+For a type parameter `T` with a `new()` constraint, `new T()` is emitted as a call to `System.Activator.CreateInstance<T>()` which ignores any explicit parameterless constructor.
+A warning is reported if there is an accessible parameterless constructor.
+
+_Should `Activator.CreateInstance()` call the parameterless constructor if available and accessible?_
+```csharp
+static T Create<T>() where T : new() => new T();
+
+x = Create<NoConstructor>();      // ok
+y = Create<PublicConstructor>();  // warning: constructor ignored
+z = Create<PrivateConstructor>(); // ok
+```
+
+A local or field of value type `S` that is not explicitly initialized is zeroed.
+The compiler already reports `error: use of unassigned local` for uninitialized locals of a value type that is not empty, and there is no additional handling beyond that for value types with explicit parameterless constructors.
+```csharp
+NoConstructor x;
+PublicConstructor y;
+PrivateConstructor z;
+x.ToString(); // error: use of unassigned local (unless type is empty)
+y.ToString(); // error: use of unassigned local (unless type is empty)
+z.ToString(); // error: use of unassigned local (unless type is empty)
+```
+
+Array allocation ignores any parameterless constructor and generates zeroed elements.
+A warning is reported if there is an accessible parameterless constructor.
+```csharp
+var a = new NoConstructor[1];      // ok
+var b = new PublicConstructor[1];  // warning: constructor ignored
+var c = new PrivateConstructor[1]; // ok
+```
+
```

Comment:
Couple of details think we need to make explicit here:

- If a `struct` has no fields and a paremeterless constructor do we consider a local of that type to be definitely assigned? 
- Can parameterless `struct` be applied to all `struct` types: `ref` and `readonly` specifically 
-